add skip tags aggregator to support TTS service spelling out tags

aconchillo opened the pull request:
#### Please describe the changes in your PR. If it is addressing an issue, please reference that as well.

This fixes and issue where emails are currently considered an end of sentence. It also add an example that uses Rime's `spell()` function to gather and confirm emails from the user.

markbackman commented on src/pipecat/utils/string.py:
```diff
@@ -17,9 +17,26 @@
+
 ENDOFSENTENCE_PATTERN = re.compile(ENDOFSENTENCE_PATTERN_STR, re.VERBOSE)
 
+EMAIL_PATTERN = re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}")
```
It looks like this pattern covers it, but I see `first.last@email.com` getting split by the TTS providers:

Cartesia:
```
Generating TTS: [ <spell>first.]
Generating TTS: [last@email.com</spell>]
```

Same for Rime.
```
Generating TTS: [ Can you please confirm:

1.  spell(first.]
Generating TTS: [last@email.com)
```

Is that an issue with the TTS provider or the parsing logic?

aconchillo replied:
It should be our parsing logic. Checking.

aconchillo replied:
Oh... maybe because it treats `<spell>` as if it was an email and it decides it's not an email.

aconchillo replied:
Complicated, solution I guess is to add a space.

aconchillo replied:
Acutally, nevermind. The problem is the LLM tokens come slowly.

filipi87 commented on src/pipecat/utils/string.py:
```diff
@@ -17,9 +18,36 @@
-    match = ENDOFSENTENCE_PATTERN.search(text.rstrip())
+    text = text.rstrip()
+
+    # Replace email dots by ampersands so we can find the end of sentence.
```
Really nit: Would be worth to add a small example in the comment, just to make easier to understand.
For example:
- this email: first.last@email.com
- will be replaced by this: first&last@email&com

aconchillo replied:
Done!

aconchillo commented:
@markbackman @filipi87 I fixed the case that Mark found. From the CHANGELOG:

```
- Added a new parameter `eos_skip_tags` to `TTSService` to prevent breaking a
  sentence prematurely. The new `eos_skip_tags` is a list of tuples where each
  tuple has a start and an end tag. For example, the tuple ("<email>",
  "</email>") would prevent the dot in "<email>foo@pipecat.ai</email>" to be
  considered an end-of-sentence. Note that Pipecat aggregates LLM tokens as fast
  as they come and provides sentences to the TTS service. Sometimes it's
  possible that the LLM, for example considering the previous example, returns
  the following tokens: "foo", then "@pipecat" and finally ".ai" which makes it
  hard to decide end-of-sentence limits.
```

markbackman commented:
@aconchillo is this working for you? With the latest changes, I see:
```
2025-03-02 08:34:07.520 | DEBUG    | pipecat.services.cartesia:run_tts:275 - CartesiaTTSService#0: Generating TTS [ Just to confirm, I have recorded the following:

1. <spell>first.]
2025-03-02 08:34:07.789 | DEBUG    | pipecat.services.cartesia:run_tts:275 - CartesiaTTSService#0: Generating TTS [lastname@gmail.com</spell>
2. <spell>f.]
2025-03-02 08:34:07.959 | DEBUG    | pipecat.services.cartesia:run_tts:275 - CartesiaTTSService#0: Generating TTS [lastname@hotmail.com</spell>

Is this correct?]
```

and:

```
2025-03-02 08:36:50.295 | DEBUG    | pipecat.services.rime:run_tts:379 - RimeHttpTTSService#0: Generating TTS [To confirm, are these your email addresses: spell(first.]
2025-03-02 08:36:50.804 | DEBUG    | pipecat.services.rime:run_tts:379 - RimeHttpTTSService#0: Generating TTS [lastname@gmail.com) and spell(f.]
2025-03-02 08:36:51.187 | DEBUG    | pipecat.services.rime:run_tts:379 - RimeHttpTTSService#0: Generating TTS [lastname@hotmail.com)?]
```

I also see Rime outputting `spell` in the TTS.

Do you see the same?

---

Also, should we use `RimeTTSService` in the 35 foundational example?

aconchillo commented:
> @aconchillo is this working for you? With the latest changes, I see:
> 
> ```
> 2025-03-02 08:34:07.520 | DEBUG    | pipecat.services.cartesia:run_tts:275 - CartesiaTTSService#0: Generating TTS [ Just to confirm, I have recorded the following:
> 
> 1. <spell>first.]
> 2025-03-02 08:34:07.789 | DEBUG    | pipecat.services.cartesia:run_tts:275 - CartesiaTTSService#0: Generating TTS [lastname@gmail.com</spell>
> 2. <spell>f.]
> 2025-03-02 08:34:07.959 | DEBUG    | pipecat.services.cartesia:run_tts:275 - CartesiaTTSService#0: Generating TTS [lastname@hotmail.com</spell>
> 
> Is this correct?]
> ```
> 
> and:
> 
> ```
> 2025-03-02 08:36:50.295 | DEBUG    | pipecat.services.rime:run_tts:379 - RimeHttpTTSService#0: Generating TTS [To confirm, are these your email addresses: spell(first.]
> 2025-03-02 08:36:50.804 | DEBUG    | pipecat.services.rime:run_tts:379 - RimeHttpTTSService#0: Generating TTS [lastname@gmail.com) and spell(f.]
> 2025-03-02 08:36:51.187 | DEBUG    | pipecat.services.rime:run_tts:379 - RimeHttpTTSService#0: Generating TTS [lastname@hotmail.com)?]
> ```
> 
> I also see Rime outputting `spell` in the TTS.
> 
> Do you see the same?

Yes, for Rime I see they return `spell(....)`. Cartesia returns individual letters and doesn't return the the spelling out tags.

> 
> Also, should we use `RimeTTSService` in the 35 foundational example?

Both Cartesia and Rime work.

markbackman commented on CHANGELOG.md:
```diff
@@ -34,6 +44,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  how to gather user emails. The example uses's Cartesia's `<spell></spell>`
+  tags and Rime `spell()` function to spell out the emails for confirmation.
+
+### Other
```
This is duplicated.

markbackman reviewed (approved):
LGTM! Just one comment about the changelog.

As you mentioned, there's an issue with Rime which they'll resolve.

markbackman reviewed (approved):
LGTM! Just two small comments about the demo.

The Cartesia example works well (but the context update is strange: `Are they F, I, R, S, T, dot. L, A, S, T, N, A, M, E, at, G, M, A, I, L, dot,`. I saw you flag this to Cartesia)

The Rime example isn't working as expected. I think this is an issue that the Rime team needs to sort out.

aconchillo commented:
@markbackman @filipi87 This is now implemented on top of the new text aggregators. I'm not sure about the name `SkipTagsAggregator`. I was tempted to follow the `PatternPairAggregator` naming, for example `SkipPairAggregator` but I thought tags was more clear. Any thoughts? Also, I haven't added a `add_pair` function, I'm not sure it's really necessary.

aconchillo commented:
> LGTM! Just two small comments about the demo.
> 
> The Cartesia example works well (but the context update is strange: `Are they F, I, R, S, T, dot. L, A, S, T, N, A, M, E, at, G, M, A, I, L, dot,`. I saw you flag this to Cartesia)
> 
> The Rime example isn't working as expected. I think this is an issue that the Rime team needs to sort out.

Btw, about this, I have asked Cartesia and I believe they will add something so we can re-normalize the spelled out words.

aconchillo commented:
Another thing that I realized is that now we support adding a list of aggregators. We set aggregators already for Cartesia and Rime. If the user wants to add their own aggregators they will need to add the pre-defined ones as well. For example, let's say they want to add the `PatternPairAggregator` and they also need `SkipTagsAggregator`. What they should do is:

```
text_aggregators = [PatternPairAggregator(...), SkipTagsAggregator(...)]
```

aconchillo commented:
OK, I changed it back to a single text aggregator. We can think about how to do multiples in the future.
